### ISSUE TITLE
Switch to recordmcount.c (CONFIG_HAVE_C_RECORDMCOUNT)

### ISSUE THREAD

**Issue by mpe** [linuxppc/issues#149]
Ftrace uses a script `recordmcount.pl` to post-process the kernel and find `mcount` locations.

There is a now a C version, `recordmcount.c` and arches should be ported over to it.

**Comment by chleroy** [comment]
What do we have to do here ? Is it just selecting HAVE_C_RECORDMCOUNT from arch/powerpc/Kconfig ?

In that case it seems to work.

**Comment by mpe** [comment]
I'm not sure.

Other arches required some changes to the C code, eg. torvalds/linux@af64d2aa872a174772ffc00e2558083f70193acb.

If the content of the mcount section is the same before and after then that's probably sufficient evidence that it works.

**Comment by chleroy** [comment]
As far as I can see, in that commit they are adding support for HAVE_FTRACE_MCOUNT_RECORD

As far as I understand, in powerpc we already have that support, but we use recordmcount.pl.

It looks like recordmcount.c already have code for powerpc:

	case EM_PPC:	 reltype = R_PPC_ADDR32;   gpfx = '_'; break;
	case EM_PPC64:	 reltype = R_PPC64_ADDR64; gpfx = '_'; break;

See linuxppc/linux@178a0600340b
Or linuxppc/linux@64575f918f32

**Comment by chleroy** [comment]
I built pmac32_defconfig with HAVE_C_RECORDMCOUNT and without. Looks like the perl version is missing several. I checked a few of the additional ones I got when building with HAVE_C_RECORDMCOUNT and they all are valid.

So I guess it is worth switching to HAVE_C_RECORDMCOUNT. I'll send a patch for it.

**Comment by chleroy** [comment]
Patch at https://patchwork.ozlabs.org/patch/1096356/

**Comment by chleroy** [comment]
When building pmac32_defconfig + CONFIG_FTRACE_MCOUNT_RECORD, I get 22071 calls to mcount.
When using HAVE_C_RECORDMCOUNT, I get 22071 items in the table:

```
c09123a4 g       .init.data	00000000 __stop_mcount_loc
c08fcac8 g       .init.data	00000000 __start_mcount_loc

```
Without HAVE_C_RECORDMCOUNT, 10 are missing

```
c0912328 g       .init.data	00000000 __stop_mcount_loc
c08fcac8 g       .init.data	00000000 __start_mcount_loc

```

The first missing one is in arch/powerpc/kernel/traps.c

Here are a few extracts from arch/powerpc/kernel/traps.o:
With HAVE_C_RECORDMCOUNT

```
 20 __mcount_loc  00000080  00000000  00000000  00005758  2**2
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, DATA

RELOCATION RECORDS FOR [__mcount_loc]:
OFFSET   TYPE              VALUE 
00000000 R_PPC_ADDR32      .text+0x00000008
00000004 R_PPC_ADDR32      .text+0x0000012c
00000008 R_PPC_ADDR32      .text+0x00000190
0000000c R_PPC_ADDR32      .text+0x000001d4
00000010 R_PPC_ADDR32      .text+0x00000214
00000014 R_PPC_ADDR32      .text+0x0000051c
00000018 R_PPC_ADDR32      .text+0x00000554
0000001c R_PPC_ADDR32      .text+0x000005e4
00000020 R_PPC_ADDR32      .text+0x0000066c
00000024 R_PPC_ADDR32      .text+0x00000694
00000028 R_PPC_ADDR32      .text+0x0000080c
0000002c R_PPC_ADDR32      .text+0x000009ec
00000030 R_PPC_ADDR32      .text+0x00000c6c
00000034 R_PPC_ADDR32      .text+0x00000ca4
00000038 R_PPC_ADDR32      .text+0x00000d0c
0000003c R_PPC_ADDR32      .text+0x00000d68
00000040 R_PPC_ADDR32      .text+0x00000e10
00000044 R_PPC_ADDR32      .text+0x00000e48
00000048 R_PPC_ADDR32      .text+0x00000efc
0000004c R_PPC_ADDR32      .text+0x00000f38
00000050 R_PPC_ADDR32      .text+0x00001794
00000054 R_PPC_ADDR32      .text+0x000017c8
00000058 R_PPC_ADDR32      .text+0x000018c4
0000005c R_PPC_ADDR32      .text+0x00001930
00000060 R_PPC_ADDR32      .text+0x00001988
00000064 R_PPC_ADDR32      .text+0x00001a00
00000068 R_PPC_ADDR32      .text+0x00001a78
0000006c R_PPC_ADDR32      .text+0x00001ac0
00000070 R_PPC_ADDR32      .text+0x00001b18
00000074 R_PPC_ADDR32      .text+0x00001c84
00000078 R_PPC_ADDR32      .text+0x00001ce0
0000007c R_PPC_ADDR32      .init.text+0x00000008

```
Without HAVE_C_RECORDMCOUNT

```
 13 __mcount_loc  0000007c  00000000  00000000  000024fc  2**0
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, DATA

RELOCATION RECORDS FOR [__mcount_loc]:
OFFSET   TYPE              VALUE 
00000000 R_PPC_ADDR32      die_will_crash+0x00000008
00000004 R_PPC_ADDR32      die_will_crash+0x0000006c
00000008 R_PPC_ADDR32      die_will_crash+0x000000b0
0000000c R_PPC_ADDR32      die_will_crash+0x000000f0
00000010 R_PPC_ADDR32      die_will_crash+0x000003f8
00000014 R_PPC_ADDR32      die_will_crash+0x00000430
00000018 R_PPC_ADDR32      die_will_crash+0x000004c0
0000001c R_PPC_ADDR32      die_will_crash+0x00000548
00000020 R_PPC_ADDR32      die_will_crash+0x00000570
00000024 R_PPC_ADDR32      die_will_crash+0x000006e8
00000028 R_PPC_ADDR32      die_will_crash+0x000008c8
0000002c R_PPC_ADDR32      die_will_crash+0x00000b48
00000030 R_PPC_ADDR32      die_will_crash+0x00000b80
00000034 R_PPC_ADDR32      die_will_crash+0x00000be8
00000038 R_PPC_ADDR32      die_will_crash+0x00000c44
0000003c R_PPC_ADDR32      die_will_crash+0x00000cec
00000040 R_PPC_ADDR32      die_will_crash+0x00000d24
00000044 R_PPC_ADDR32      die_will_crash+0x00000dd8
00000048 R_PPC_ADDR32      die_will_crash+0x00000e14
0000004c R_PPC_ADDR32      die_will_crash+0x00001670
00000050 R_PPC_ADDR32      die_will_crash+0x000016a4
00000054 R_PPC_ADDR32      die_will_crash+0x000017a0
00000058 R_PPC_ADDR32      die_will_crash+0x0000180c
0000005c R_PPC_ADDR32      die_will_crash+0x00001864
00000060 R_PPC_ADDR32      die_will_crash+0x000018dc
00000064 R_PPC_ADDR32      die_will_crash+0x00001954
00000068 R_PPC_ADDR32      die_will_crash+0x0000199c
0000006c R_PPC_ADDR32      die_will_crash+0x000019f4
00000070 R_PPC_ADDR32      die_will_crash+0x00001b60
00000074 R_PPC_ADDR32      die_will_crash+0x00001bbc
00000078 R_PPC_ADDR32      trap_init+0x00000008

```

**Comment by chleroy** [comment]
Missign one is there:

```
00000000 <exception_common.part.3>:
       0:	7c 08 02 a6 	mflr    r0
       4:	90 01 00 04 	stw     r0,4(r1)
       8:	48 00 00 01 	bl      8 <exception_common.part.3+0x8>
			8: R_PPC_REL24	_mcount
       c:	3d 20 00 00 	lis     r9,0
			e: R_PPC_ADDR16_HA	show_unhandled_signals
      10:	94 21 ff d0 	stwu    r1,-48(r1)
      14:	81 29 00 00 	lwz     r9,0(r9)
			16: R_PPC_ADDR16_LO	show_unhandled_signals
      18:	7c 08 02 a6 	mflr    r0
      1c:	2f 89 00 00 	cmpwi   cr7,r9,0
      20:	bf 61 00 1c 	stmw    r27,28(r1)

```

**Comment by chleroy** [comment]
Second missign one is:

```
c001d638 <clone_property.isra.2>:
c001d638:	7c 08 02 a6 	mflr    r0
c001d63c:	90 01 00 04 	stw     r0,4(r1)
c001d640:	4b ff d8 4d 	bl      c001ae8c <_mcount>
c001d644:	94 21 ff e0 	stwu    r1,-32(r1)
c001d648:	3d 20 c0 87 	lis     r9,-16249
```

Wonder if it doesn't have something to do with objects having a dot in their name.

**Comment by mpe** [comment]
I see something similar with ppc64le_defconfig. The C version finds 37 extra entries.

Of those 30 have a `.` in the name:
```
tg3_restart_hw.part.54+0xc
clone_property.isra.1+0xc
e1000_mng_enable_host_if.isra.1+0xc
xfs_dir2_data_freeremove.isra.0+0x4
lmb_is_removable.isra.3.part.4+0xc
earlycon_init.constprop.1+0xc
xfs_dir3_leaf_log_bests.isra.2+0xc
__earlyonly_bootmem_alloc.constprop.1+0xc
mce_flush.part.1+0xc
ext2_xattr_cache_insert.isra.0+0xc
_mpic_map_mmio.isra.3+0xc
copy_vsx_to_user.part.1+0x4
do_sysctl.isra.1.part.2+0xc
nfs_parse_server_name.constprop.2+0xc
gssx_enc_buffer.isra.3+0xc
cpuidle_switch_governor.part.0+0xc
vio_bus_scan_register_devices.part.2+0xc
sched_numa_warn.part.3+0xc
__func_get_name.constprop.2+0xc
restore_sigcontext.constprop.1+0xc
__parse_crashkernel.constprop.0+0xc
lmb_to_memblock.isra.1+0xc
utimes_common.isra.0+0xc
tg3_warn_mgmt_link_flap.part.43+0xc
process_vm_rw_core.isra.0+0xc
xfs_dir3_leaf_find_stale.isra.0+0x4
xfs_dir3_leaf_log_tail.isra.1+0xc
snoop_urb.part.2+0xc
group_adjust_blocks.isra.3.part.4+0xc
key_gc_unused_keys.constprop.1+0xc
```

That seems like it's explained by `.` being missing in `$function_regex`.

Then there's a few that are all `__weak` functions:
```
arch_kexec_apply_relocations+0xc
elf_core_extra_phdrs+0x4
elf_core_write_extra_phdrs+0x4
elf_core_write_extra_data+0x4
elf_core_extra_data_size+0x4
abort+0x4
arch_kexec_apply_relocations_add+0xc
```

Not sure why they're not being picked up.

**Comment by mpe** [comment]
This catches most of them:
```
diff --git a/scripts/recordmcount.pl b/scripts/recordmcount.pl
index d24759214efd..3d1e4a2257b9 100755
--- a/scripts/recordmcount.pl
+++ b/scripts/recordmcount.pl
@@ -275,7 +275,7 @@ if ($arch eq "x86_64") {

     $local_regex = "^[0-9a-fA-F]+\\s+t\\s+(\\.?\\S+)";
     # See comment in the sparc64 section for why we use '\w'.
-    $function_regex = "^([0-9a-fA-F]+)\\s+<(\\.?\\w*?)>:";
+    $function_regex = "^([0-9a-fA-F]+)\\s+<(\\.?[\\w\\.]*?)>:";
     $mcount_regex = "^\\s*([0-9a-fA-F]+):.*\\s\\.?_mcount\$";

     if ($endian eq "big") {
```

**Comment by chleroy** [comment]
See linuxppc/linux@f079bb3c5f2978b2c1a13098ab2a8c32e5d1ee3d

**Comment by mpe** [comment]
Merged as torvalds/linux@f079bb3c5f2978b2c1a13098ab2a8c32e5d1ee3d.

Thanks @chleroy :champagne:
